### PR TITLE
Prepare npc, spell, character, and item for new ai

### PR DESCRIPTION
**Pull request by KorGgenT**
#### Summary
Infrastructure "Prepare npc, spell, character, and item for new ai"

#### Purpose of change
These are all changes that were necessary to decouple some things so i could write the new npc_attack class. 

#### Describe the solution

- changed item::effective_dps over to Creature instead of monster
- added a (very) basic spell::dps()
- made a few things const that could/should be
- pulled sound volume out of spell::make_sound and made it its own function
- updated npc::attitude_to to use the ai_cache
- pulled a few parts out of the throw function into their own functions for readability and separation

#### Describe alternatives you've considered
including these changes in the npc ai pr i suppose

#### Testing
the testing i've done so far is after the npc_attack code, and i haven't seen any oddities. this code should not change any behavior.

#### Additional context
This is a prerequisite PR for npc_attack.

### CODE REVIEW

File: src/activity_handlers.cpp [hunk 1]
```diff
@@ -4213,25 +4213,47 @@ void activity_handlers::spellcasting_finish( player_activity *act, player *p )
...
+                const tripoint local_target = get_map().getlocal( act->coords.front() );
+                if( spell_being_cast.is_valid_target( *p, local_target ) ) {
+                    target = local_target;
+                } else {
+                    npc_attack_spell npc_spell( spell_being_cast.id() );
+                    // recalculate effectiveness because it's been a few turns since the npc started casting.
+                    const npc_attack_effectiveness effectiveness = npc_spell.evaluate( *p->as_npc(),
+                            p->last_target.lock().get() );
+                    if( effectiveness < 0 ) {
+                        add_msg_debug( "%s cancels casting %s, target lost" );
```
Comment (BrettDong) [inline src/activity_handlers.cpp:4249]: Is parameter missing here?

**Comment by KorGgenT** [comment]
uh oh, i messed up git add -p. it has some stray npc_attack stuff in here.

**Comment by KorGgenT** [comment]
all right i just moved the activity changes to the other branch instead of here.

File: src/ranged.cpp [hunk 2]
```diff
@@ -961,64 +963,112 @@ int Character::throwing_dispersion( const item &to_throw, Creature *critter,
...
+
+int Character::thrown_item_total_damage_raw( const item &thrown ) const
+{
+    projectile proj = thrown_item_projectile( thrown );
+    const units::volume volume = thrown.volume();
+    proj.impact.add_damage( damage_type::BASH, std::min( thrown.weight() / 100.0_gram,
+                            static_cast<double>( thrown_item_adjusted_damage( thrown ) ) ) );
+    // Item will shatter upon landing, destroying the item, dealing damage, and making noise
+    if( !thrown.active && thrown.made_of( material_id( "glass" ) ) &&
+        rng( 0, units::to_milliliter( 2_liter - volume ) ) < get_str() * 100 ) {
```
Comment (BrettDong) [inline src/ranged.cpp:1026]: Is it possible that `volume` is greater than 2 litres?

File: src/npc.cpp [hunk 3]
```diff
@@ -2135,6 +2135,26 @@ bool npc::is_travelling() const
 
 Creature::Attitude npc::attitude_to( const Creature &other ) const
 {
+    const auto same_as = []( const Creature * lhs, const Creature * rhs ) {
+        return &lhs == &rhs;
+    };
+
+    for( const weak_ptr_fast<Creature> buddy : ai_cache.friends ) {
```
Comment (BrettDong) [inline src/npc.cpp:2142]:
```suggestion
    for( const weak_ptr_fast<Creature> &buddy : ai_cache.friends ) {
```

File: src/npc.cpp [hunk 4]
```diff
@@ -2135,6 +2135,26 @@ bool npc::is_travelling() const
...
+        if( same_as( &other, buddy.lock().get() ) ) {
+            return Creature::Attitude::FRIENDLY;
+        }
+    }
+    for( const weak_ptr_fast<Creature> enemy : ai_cache.hostile_guys ) {
```
Comment (BrettDong) [inline src/npc.cpp:2147]:
```suggestion
    for( const weak_ptr_fast<Creature> &enemy : ai_cache.hostile_guys ) {
```

File: src/npc.cpp [hunk 5]
```diff
@@ -2135,6 +2135,26 @@ bool npc::is_travelling() const
...
+        if( same_as( &other, enemy.lock().get() ) ) {
+            return Creature::Attitude::HOSTILE;
+        }
+    }
+    for( const weak_ptr_fast<Creature> neutral : ai_cache.neutral_guys ) {
```
Comment (BrettDong) [inline src/npc.cpp:2152]:
```suggestion
    for( const weak_ptr_fast<Creature> &neutral : ai_cache.neutral_guys ) {
```

**Comment by BrettDong** [comment]
I'm getting linking error
```
Undefined symbols for architecture arm64:
  "npc_short_term_cache::closest_enemy_to_friendly_distance() const", referenced from:
      npc::closest_enemy_to_friendly_distance() const in npc.o
ld: symbol(s) not found for architecture arm64
```
Is it because you omitted including `npc_short_term_cache::closest_enemy_to_friendly_distance()` in this PR?

File: src/npcmove.cpp [hunk 6]
```diff
@@ -369,6 +369,20 @@ static bool too_close( const tripoint &critter_pos, const tripoint &ally_pos, co
     return rl_dist( critter_pos, ally_pos ) <= def_radius;
 }
 
+int npc_short_term_cache::closest_enemy_to_friendly_distance() const
+{
+    int distance = INT_MAX;
+    for( const weak_ptr_fast<Creature> buddy : friends ) {
+        for( const weak_ptr_fast<Creature> enemy : hostile_guys ) {
+            if( buddy.expired() || enemy.expired() ) {
```
Comment (BrettDong) [inline src/npcmove.cpp:377]: `buddy.expired()` check can be moved to the outer loop.

File: src/ranged.cpp (hunk 2 above)
Comment (KorGgenT) [reply src/ranged.cpp:1026]: i guess so but this isn't new code

File: src/npcmove.cpp [hunk 7]
```diff
@@ -369,6 +369,20 @@ static bool too_close( const tripoint &critter_pos, const tripoint &ally_pos, co
     return rl_dist( critter_pos, ally_pos ) <= def_radius;
 }
 
+int npc_short_term_cache::closest_enemy_to_friendly_distance() const
+{
+    int distance = INT_MAX;
+    for( const weak_ptr_fast<Creature> buddy : friends ) {
```
Comment (BrettDong) [inline src/npcmove.cpp:375]:
```suggestion
    for( const weak_ptr_fast<Creature> &buddy : friends ) {
```

File: src/npcmove.cpp [hunk 8]
```diff
@@ -369,6 +369,20 @@ static bool too_close( const tripoint &critter_pos, const tripoint &ally_pos, co
...
+        for( const weak_ptr_fast<Creature> enemy : hostile_guys ) {
```
Comment (BrettDong) [inline src/npcmove.cpp:376]:
```suggestion
        for( const weak_ptr_fast<Creature> &enemy : hostile_guys ) {
```

**Pull request merged**